configfile: "scripts/config/config.yaml",

##### load rules #####
rule all:
    input:
        expand("results/fastqc_init/{sample}_fastqc.zip", sample = config["samples"]),
        expand("results/fastqc_init/{sample}_fastqc.html", sample = config["samples"]),
        expand("results/trim/{subsample}_1_trim_paired.fastq.gz", subsample = config["subsamples"]),
        expand("results/trim/{subsample}_1_trim_unpaired.fastq.gz", subsample = config["subsamples"]),
        expand("results/trim/{subsample}_2_trim_paired.fastq.gz", subsample = config["subsamples"]),
        expand("results/trim/{subsample}_2_trim_unpaired.fastq.gz", subsample = config["subsamples"]),
        expand("results/fastqc_post/{sample}_trim_paired_fastqc.zip", sample = config["samples"]),
        expand("results/fastqc_post/{sample}_trim_paired_fastqc.html", sample = config["samples"]),
        expand("results/bowtie2/{subsample}_trim_mapped_sorted_q2.bam", subsample = config["subsamples"]),
        expand("results/bowtie2/{subsample}_trim_mapped_sorted_q2.bam.bai", subsample = config["subsamples"]),
	expand("results/bowtie2/{subsample}.log", subsample = config["subsamples"]),
        expand("results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam", subsample = config["subsamples"]),
        expand("results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam.bai", subsample = config["subsamples"]),
        expand("results/picard/{subsample}_trim_mapped_sorted_q2_dups.txt", subsample = config["subsamples"]),
        expand("results/picard/{subsample}_trim_mapped_sorted_q2.log", subsample = config["subsamples"]),
        expand("results/deeptools/{subsample}_multiBamSum.npz", subsample = config["multibam"]),
        expand("results/deeptools/{subsample}_heatmap-pearson.pdf", subsample = config["multibam"]),
        expand("results/deeptools/{subsample}_heatmap-pearson.tab", subsample = config["multibam"]),
        expand("results/deeptools/{subsample}_PCA.png", subsample = config["multibam"]),
        expand("results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_model.r", subsample = config["subsamples"]),
        expand("results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_peaks.narrowPeak", subsample = config["subsamples"]),
        expand("results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_peaks.xls", subsample = config["subsamples"]),
        expand("results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_summits.bed", subsample = config["subsamples"])

rule fastqc_init:
    input:
        "data/subset/{sample}.fastq.gz"
    output:
        html = "results/fastqc_init/{sample}_fastqc.html",
        zip = "results/fastqc_init/{sample}_fastqc.zip"
    conda:
        "envs/qc.yaml"
    threads: 2
    log:
        "log/fastqc_init_{sample}.log"
    message:
        "Initial Quality Control (qc): {input}"
    shell:
        """
        mkdir -p results/fastqc_init
        /home/ubuntu/atacseq/fastqc-0.11.7/fastqc {input} -o "results/fastqc_init" -t {threads} 2>{log}
        """

rule trim:
    input:
        fwd = "data/subset/{subsample}_1.fastq.gz",
        rev = "data/subset/{subsample}_2.fastq.gz"
    output:
        fwd_paired = "results/trim/{subsample}_1_trim_paired.fastq.gz",
        fwd_unpaired = "results/trim/{subsample}_1_trim_unpaired.fastq.gz",
        rev_paired = "results/trim/{subsample}_2_trim_paired.fastq.gz",
        rev_unpaired = "results/trim/{subsample}_2_trim_unpaired.fastq.gz"
    conda:
        "envs/trim.yaml"
    threads: 6
    log: 
        "log/trim_{subsample}.log"
    message: 
        "Trimming (trim) : {input}"
    shell:
        """
        mkdir -p results/trim
        java -jar /home/ubuntu/atacseq/trimmomatic-0.38/trimmomatic-0.38.jar PE -threads {threads} \
            -trimlog results/trim.log -summary results/trim/stats \
            {input.fwd} {input.rev} \
            {output.fwd_paired} {output.fwd_unpaired} {output.rev_paired} {output.rev_unpaired} \
                ILLUMINACLIP:data/NexteraPE-PE.fa:2:30:10:2:keepBothReads \
                LEADING:3 \
                TRAILING:3 \
                SLIDINGWINDOW:4:15 \
                MINLEN:33
        """

rule fastqc_post:
    input:
        "results/trim/{sample}_trim_paired.fastq.gz"
    output:
        html = "results/fastqc_post/{sample}_trim_paired_fastqc.html",
        zip = "results/fastqc_post/{sample}_trim_paired_fastqc.zip"
    conda:
        "envs/qc.yaml"
    threads: 2
    log:
        "log/fastqc_post_{sample}.log"
    message:
        "Trimming and  quality control : {input}"
    shell:
        """
        mkdir -p results/fastqc_post
        /home/ubuntu/atacseq/fastqc-0.11.7/fastqc {input} -o "results/fastqc_post" -t {threads} 2>{log}
        """

rule bowtie2:
    input:
        fwd = "results/trim/{subsample}_1_trim_paired.fastq.gz",
        rev = "results/trim/{subsample}_2_trim_paired.fastq.gz",
        idx_bt2 = "/home/ubuntu/atacseq/data/Mus_musculus_GRCm39/bowtie2"
    output:
        bam = "results/bowtie2/{subsample}_trim_mapped_sorted_q2.bam",
        bai = "results/bowtie2/{subsample}_trim_mapped_sorted_q2.bam.bai",
        log = "results/bowtie2/{subsample}.log"
    conda:
        "envs/bowtie2.yaml"
    threads: 8
    log:
        "log/bowtie2_{subsample}.log"
    message:
        "Mapping in reference genome : {input}"
    shell:
        """
        mkdir -p results/bowtie2
        /home/ubuntu/atacseq/bowtie2-2.3.4.3/bowtie2 --very-sensitive -p 8 -k 10 -x {input.idx_bt2}/all \
        -1 {input.fwd} -2 {input.rev} | samtools view -q 2 -bS - | samtools sort - -o {output.bam}
        samtools index -b {output.bam}
        samtools idxstats {output.bam} >{output.log}
        """

rule picard: 
     input:
        "results/bowtie2/{subsample}_trim_mapped_sorted_q2.bam"
     output:
        bam = "results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam",
        bai = "results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam.bai",
        txt = "results/picard/{subsample}_trim_mapped_sorted_q2_dups.txt",
        log = "results/picard/{subsample}_trim_mapped_sorted_q2.log"
     conda:
        "envs/picard.yaml"
     threads: 6
     log:
        "log/picard_{subsample}.log"
     message:
        "Mapping and clean : {input}"
     shell:
        """
        mkdir -p results/picard
        java -jar /home/ubuntu/atacseq/picard-2.18.25/picard.jar MarkDuplicates I="{input}" O={output.bam} \
        M={output.txt} REMOVE_DUPLICATES=true
        samtools index -b {output.bam}
        samtools idxstats {output.bam} > {output.log}
        """

rule deeptools:
    input:
        bam = expand("results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam", subsample = config["subsamples"])
    output:
        "results/deeptools/{subsample}_multiBamSum.npz"
    conda: 
        "envs/deeptools.yaml"
    threads: 2
    log:
        "log/deeptools_multiBamSum_{subsample}.log"
    message:
        Data visualisation first part: {input}"
    shell:
        """
        mkdir -p results/deeptools
        multiBamSummary bins --bamfiles {input.bam} -o {output}
        """

rule deeptools2:
    input:
        "results/deeptools/{subsample}_multiBamSum.npz"
    output:
        pdf = "results/deeptools/{subsample}_heatmap-pearson.pdf",
        tab = "results/deeptools/{subsample}_heatmap-pearson.tab",
        png = "results/deeptools/{subsample}_PCA.png"
    conda:
        "envs/deeptools.yaml"
    threads: 2
    log:
        "log/deeptools_corr_{subsample}.log"
    message:
        "Visualisation {input} in a correlation matrix"
    shell:
        """
        plotCorrelation \
            -in {input} \
            --corMethod pearson --skipZeros \
            --plotTitle "Pearson Correlation of Read coverage" \
            --whatToPlot heatmap --colorMap RdYlBu --plotNumbers \
            -o {output.pdf}   \
            --outFileCorMatrix {output.tab}
        plotPCA \
            -in {input} \
            -T "Read coverage PCA" \
            -o {output.png}
        """

rule macs2:
    input:
        "results/picard/{subsample}_trim_mapped_sorted_q2_nodup.bam"
    output:
        r = "results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_model.r",
        nar = "results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_peaks.narrowPeak",
        xls = "results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_peaks.xls",
        bed = "results/MACS2/{subsample}_trim_mapped_sorted_q2_nodup_summits.bed"
    conda:
        "envs/macs2.yaml"
    threads: 2
    log:
        "log/macs2_{subsample}.log"
    message:
        "Executing first data visualisation: {input}"
    shell:
        """
        mkdir -p results/MACS2
        macs2 callpeak -t {input} -n {wildcards.subsample}_trim_mapped_sorted_q2_nodup -f BAM -g mm --outdir "results/MACS2"
        """
